src/java/org/apache/cassandra/index/sai/plan/VectorTopKProcessor.java

```
@@ -0,0 +1,271 @@
+import org.apache.cassandra.db.filter.RowFilter;
+import org.apache.cassandra.db.partitions.PartitionIterator;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.db.rows.RowIterator;
+import org.apache.cassandra.index.Index;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.cassandra.utils.Pair;
+
+import static org.apache.cassandra.index.sai.disk.v1.IndexWriterConfig.MAX_TOP_K;
+
+/**
+ * Processor that scans all rows from reconciled query result and selects rows with top-k scores based on vector indexes.
+ *
+ * This processor performs the following steps:
+ * - collect rows with score into PriorityQueue that sorts rows based on score. If there are multiple vector indexes,
+ *   the final score is the sum of all vector index scores.
+ * - remove rows with the lowest scores from PQ if PQ size exceeds limit
+ * - return rows with the highest scores from PQ in token/clustering order to client
+ *
+ * Note that recall will be lower with paging, because:
+ * - page size is used as limit
+ * - for the first query, coordinator returns global top page-size rows within entire ring
+ * - for the subsequent queries, coordinators returns global top page-size rows withom range from last-returned-row to max token
+ */
+public class VectorTopKProcessor
+{
+    private final ReadCommand command;
+    private final StorageAttachedIndexQueryPlan queryPlan;
+
+    private final int limit;
+
+    public VectorTopKProcessor(ReadCommand command, StorageAttachedIndexQueryPlan queryPlan)
+    {
+        this.command = command;
+        this.queryPlan = queryPlan;
+        // avoid OOM when large limit is provided by user
+        this.limit = Math.min(command.limits().count(), MAX_TOP_K);
```

Seems to me that we should fail these queries early rather than silently change the limit. This might be confusing for users.

---

good idea. Implemented `Index#validate` to fail query if limit is unlimited or exceeds MAX_TOP_K https://github.com/datastax/cassandra/pull/613/commits/d7eb71ff76c75253cc98afc06afa81de38b28173